### PR TITLE
vim-patch:8.1.{14,288,345,410,434,438,455}

### PR DESCRIPTION
**Pull request by janlazo**

### CODE REVIEW

File: src/nvim/quickfix.c [hunk 1]
```diff
@@ -892,19 +965,20 @@ static int
...
     linenr_T lnumlast,                      // last line number to use
     char_u *qf_title,
     char_u *enc
 )
+  FUNC_ATTR_NONNULL_ARG(1)
 {
-  qfstate_T state;
-  qffields_T fields;
+  qfstate_T state = { 0 };
+  qffields_T fields = { 0 };
```
Comment (janlazo) [inline src/nvim/quickfix.c]: No `Wmissing-field-initializer` warning this time unlike https://github.com/neovim/neovim/pull/9994.

File: src/nvim/quickfix.c [hunk 2]
```diff
@@ -880,6 +880,79 @@ static int qf_parse_line(qf_info_T *qi, int qf_idx, char_u *linebuf,
   return QF_OK;
 }
 
+// Allocate the fields used for parsing lines and populating a quickfix list.
+static void qf_alloc_fields(qffields_T *pfields)
```
Comment (janlazo) [inline src/nvim/quickfix.c]: Diverged from upstream because this can't "fail".

**Comment by justinmk** [comment]
freebsd CI failure:

```
[----------] Running tests from test/functional/legacy/096_location_list_spec.lua
[ RUN      ] location list is working: -- Tests exited non-zero: SIGPIPE
```

**Comment by janlazo** [comment]
That test is flaky on bsd builds. I've seen it fail with openbsd and pass on freebsd and vice versa.

**Comment by janlazo** [comment]
@blueyed Do you see any issues with deleting `096_location_list_spec.lua`? `test_quickfix.vim` runs the same test.

**Comment by justinmk** [comment]
Seems to fail more reliably after this PR. Not sure we should delete the Lua test for that reason... I ssh'd into the freebsd CI machine, got a backtrace:

```
(lldb) bt
* thread #1, name = 'nvim', stop reason = signal SIGSEGV
  * frame #0: 0x000000000049da2e nvim`copy_loclist_stack + 430
    frame #1: 0x000000000055bed5 nvim`win_init + 277
    frame #2: 0x000000000055d95f nvim`win_split_ins + 1711
    frame #3: 0x00000000003dcc3c nvim`ex_splitview + 300
    frame #4: 0x00000000003d8552 nvim`do_one_cmd + 16194
    frame #5: 0x00000000003d382c nvim`do_cmdline + 2508
    frame #6: 0x000000000034ce6a nvim`nvim_command + 26
    frame #7: 0x000000000033ec0b nvim`handle_nvim_command + 75
    frame #8: 0x000000000045fd20 nvim`request_event + 80
    frame #9: 0x00000000003bdc49 nvim`multiqueue_process_events + 73
    frame #10: 0x000000000046f166 nvim`nv_event + 38
    frame #11: 0x00000000004688cf nvim`normal_execute + 5535    frame #12: 0x000000000051d9af nvim`state_enter + 175    frame #13: 0x000000000046225c nvim`normal_enter + 188    frame #14: 0x00000000004342e9 nvim`main + 8937    frame #15: 0x000000000033011b nvim`_start(ap=<unavailable>, cleanup=<unavailable>) at crt1.c:76
```

I will try to build with debug info and get a better trace.

**Comment by blueyed** [comment]
> @blueyed Do you see any issues with deleting `096_location_list_spec.lua`? `test_quickfix.vim` runs the same test.

I'm :+1: on going with Vim's/upstream tests.

**Comment by justinmk** [comment]
```
(lldb) bt
* thread #1, name = 'nvim', stop reason = signal SIGSEGV
  * frame #0: 0x00000000004a5d3e nvim`copy_loclist_stack [inlined] copy_loclist_entries at quickfix.c:1659
    frame #1: 0x00000000004a5cb7 nvim`copy_loclist_stack [inlined] copy_loclist at quickfix.c:1696
    frame #2: 0x00000000004a5c14 nvim`copy_loclist_stack(from=<unavailable>, to=<unavailable>) at quickfix.c:1745
    frame #3: 0x0000000000565c05 nvim`win_init(newp=0x0000000800cfb800, oldp=0x0000000800cf4000, flags=2) at window.c:1534
    frame #4: 0x000000000056769f nvim`win_split_ins(size=0, flags=<unavailable>, new_wp=0x0000000000000000, dir=0) at window.c:1280
    frame #5: 0x00000000003e399c nvim`ex_splitview(eap=0x00007fffffffd860) at ex_docmd.c:6792
    frame #6: 0x00000000003df2b2 nvim`do_one_cmd(cmdlinep=0x00007fffffffdb50, flags=10, cstack=<unavailable>, fgetline=0x0000000000000000, cookie=0x0000000000000000) at ex_docmd.c:2274
    frame #7: 0x00000000003da58c nvim`do_cmdline(cmdline=<unavailable>, fgetline=<unavailable>, cookie=0x0000000000000000, flags=10) at ex_docmd.c:596
    frame #8: 0x000000000035239a nvim`nvim_command(command=<unavailable>, err=0x00007fffffffe140) at vim.c:98
    frame #9: 0x0000000000343c0b nvim`handle_nvim_command(channel_id=<unavailable>, args=Array @ 0x00007fffffffe100, error=<unavailable>) at dispatch_wrappers.generated.h:2467
    frame #10: 0x00000000004676d0 nvim`request_event(argv=<unavailable>) at channel.c:382
    frame #11: 0x00000000003c45b9 nvim`multiqueue_process_events(this=0x0000000800c70060) at multiqueue.c:157
    frame #12: 0x0000000000476d56 nvim`nv_event(cap=0x00007fffffffe328) at normal.c:8019
    frame #13: 0x0000000000470396 nvim`normal_execute(state=0x00007fffffffe2a0, key=<unavailable>) at normal.c:1131
    frame #14: 0x000000000052671f nvim`state_enter(s=0x00007fffffffe2a0) at state.c:69
    frame #15: 0x0000000000469c0c nvim`normal_enter(cmdwin=<unavailable>, noexmode=<unavailable>) at normal.c:463
    frame #16: 0x000000000043bb69 nvim`main(argc=1, argv=<unavailable>) at main.c:576
    frame #17: 0x000000000033511b nvim`_start(ap=<unavailable>, cleanup=<unavailable>) at crt1.c:76
```

File: src/nvim/quickfix.c [hunk 3]
```diff
@@ -1583,116 +1622,134 @@ static qf_info_T *ll_get_or_alloc_list(win_T *wp)
...
+      to_qfl->qf_ptr = prevp;  // current location
+    }
+  }
+
+  return OK;
+}
+
+// Copy the specified location list 'from_qfl' to 'to_qfl'.
+static int copy_loclist(const qf_list_T *restrict from_qfl,
+                        qf_list_T *restrict to_qfl,
```
Comment (janlazo) [inline src/nvim/quickfix.c]: Does it work after removing `restrct`?

Comment (justinmk) [reply src/nvim/quickfix.c]: will try.

Comment (justinmk) [reply src/nvim/quickfix.c]: yes, test passes, no crash. I removed all `restrict` cases in quickfix.c. Why are we adding that?

**Comment by justinmk** [comment]
> The restrict keyword is a declaration of intent given by the programmer to the compiler. It says that for the lifetime of the pointer, only the pointer itself or a value directly derived from it (such as pointer + 1) will be used to access the object to which it points. This limits the effects of pointer aliasing, aiding optimizations. If the declaration of intent is not followed and the object is accessed by an independent pointer, this will result in undefined behavior.

We should not be using `restrict`. Micro-managing pointer lifetime guarantees is too much cognitive effort and is unlikely to make any noticeable performance difference.

File: src/nvim/quickfix.c (hunk 3 above)
Comment (janlazo) [reply src/nvim/quickfix.c]: Optimizations :| . I thought there's no risks if it passed on the ASAN job.

Comment (janlazo) [reply src/nvim/quickfix.c]: Does it pass if `restrict` is kept for primitve types like `char`?

Comment (justinmk) [reply src/nvim/quickfix.c]: > Does it pass if `restrict` is kept for primitve types like `char`?

I would rather just avoid it entirely. Compilers don't really need these explicit hints, and it adds a lot of work for us to mentally verify these hints.

**Comment by janlazo** [comment]
Rebased to remove `restrict` on non-primitive types.

**Comment by justinmk** [comment]
travis failed on the treesitter problem, would you mind rebasing?

**Pull request merged**